src/entries/smt.rs

```diff
@@ -0,0 +1,188 @@
+    );
+
+    debug!("Define history leaves: {}", defines.len());
+    for define_db in defines {
+        let DefineDb {
+            cota_id,
+            total,
+            issued,
+            configure,
+        } = define_db;
+        let (_, key) = generate_define_key(cota_id);
+        let (_, value) =
+            generate_define_value(total.to_be_bytes(), issued.to_be_bytes(), configure);
+        smt.update(key, value).expect("SMT update leave error");
+    }
+    diff_time(start_time, "Push define history leaves to smt");
+
+    let start_time = Local::now().timestamp_millis();
+    debug!("Hold history leaves: {}", holds.len());
+    for hold_db in holds {
+        let HoldDb {
+            cota_id,
+            token_index,
+            configure,
+            state,
+            characteristic,
+        } = hold_db;
+        let (_, key) = generate_hold_key(cota_id, token_index);
+        let (_, value) = generate_hold_value(configure, state, characteristic);
+        smt.update(key, value).expect("SMT update leave error");
+    }
+    let mut withdrawal_map: HashMap<Vec<u8>, u8> = HashMap::new();
+    for withdrawal_db in withdrawals {
+        let WithdrawDb {
+            cota_id,
+            token_index,
+            configure,
+            state,
+            characteristic,
+            receiver_lock_script,
+            out_point,
+            version,
+        } = withdrawal_db;
+        let (key, value) = if version == 0 {
+            (
+                generate_withdrawal_key(cota_id, token_index).1,
+                generate_withdrawal_value(
+                    configure,
+                    state,
+                    characteristic,
+                    receiver_lock_script,
+                    out_point,
+                )
+                .1,
+            )
+        } else {
+            (
+                generate_withdrawal_key_v1(cota_id, token_index, out_point).1,
+                generate_withdrawal_value_v1(
+                    configure,
+                    state,
+                    characteristic,
+                    receiver_lock_script,
+                )
+                .1,
+            )
+        };
+        withdrawal_map.insert(generate_cota_index(cota_id, token_index), version);
+        smt.update(key, value).expect("SMT update leave error");
+    }
+    for claim_db in claims {
+        let ClaimDb {
+            cota_id,
+            token_index,
+            out_point,
+        } = claim_db;
+        let version = withdrawal_map
+            .get(&*generate_cota_index(cota_id, token_index))
+            .cloned()
+            .unwrap_or_default();
+        let (_, key) = generate_claim_key(cota_id, token_index, out_point);
+        let (_, value) = generate_claim_value(version);
+        smt.update(key, value).expect("SMT update leave error");
+    }
+    diff_time(start_time, "Push claim history leaves to smt");
+    Ok(smt)
+}
+
+pub fn save_smt_root_and_leaves(
```

Comment:
The msg is only used as a debug tag, it's better to extract it to another helper fn instead using parameter here. and it is better to move this method to `SMTStore` , we can pass an empty slice here, no need to use `Option<Vec>`:
```
  SMTStore#save_root_and_leaves(leaves: &[(H256, H256)])
```

Reply:
Good idea.